File: java/t/rendering.java
import com.iafan.plurr.*;

class RenderingTest {

  private void pass_or_fail(String n, String reference, String result) {
    System.out.printf("%s: ", n);
    if (reference.equals(result)) {
      System.out.printf("pass\n");
    } else {
      System.out.printf("fail: [%s] vs [%s]\n", result, reference);
    }
  }

  private void t (String n, Plurr p, String s, String message, String exception, String... arg) {
    String result;

    try {
      result = p.format(s, arg);
      this.pass_or_fail(n, message, result);
    } catch (PlurrInternalException e) {
      this.pass_or_fail(n, exception, "PlurrInternalException: " + e.getMessage());
    } catch (PlurrSyntaxException e) {
      this.pass_or_fail(n, exception, "PlurrSyntaxException: " + e.getMessage());
    }
  }

  public void run() throws PlurrLocaleNotFoundException {
    Plurr p = new Plurr();
    String s;

    //this.t("e.1", p, "", undefined, undefined, "", "'params' is not a hash"); // not applicable in Java version
    this.t("e.2", p, "err {", "", "PlurrSyntaxException: Unmatched { found");
    this.t("e.3", p, "err }", "", "PlurrSyntaxException: Unmatched } found");
    this.t("e.4", p, "{foo}", "", "PlurrSyntaxException: 'foo' not defined");
    this.t("e.5", p, "{N_PLURAL}", "", "PlurrSyntaxException: Value of 'N' is not a zero or positive integer number", "N", "NaN");
    this.t("e.6", p, "{N_PLURAL}", "", "PlurrSyntaxException: Value of 'N' is not a zero or positive integer number", "N", "1.5");

    s = "Do you want to delete {N_PLURAL:this {N} file|these {N} files} permanently?";

    this.t("1.1", p, s, "Do you want to delete this 1 file permanently?", "", "N", "1");
    this.t("1.2", p, s, "Do you want to delete these 5 files permanently?", "", "N", "5");

    s = "Do you want to drink {CHOICE:coffee|tea|juice}?";

    this.t("2.1", p, s, "Do you want to drink coffee?", "", "CHOICE", "0");
    this.t("2.2", p, s, "Do you want to drink tea?", "", "CHOICE", "1");
    this.t("2.3", p, s, "Do you want to drink juice?", "", "CHOICE", "2");

    p.setLocale("ru");
    s = "Удалить {N_PLURAL:этот {N} файл|эти {N} файла|эти {N} файлов} навсегда?";

    this.t("3.1", p, s, "Удалить этот 1 файл навсегда?", "", "N", "1");
    this.t("3.2", p, s, "Удалить эти 2 файла навсегда?", "", "N", "2");
    this.t("3.3", p, s, "Удалить эти 5 файлов навсегда?", "", "N", "5");
    p.setLocale("en");
    this.t("3.4", p, s, "Удалить эти 5 файла навсегда?", "","N", "5");

    s = "{X_PLURAL:{X:|One|{X}} file|{X:No|{X}} files} found.";

    this.t("4.1", p, s, "No files found.", "", "X", "0");
    this.t("4.2", p, s, "One file found.", "", "X", "1");
    this.t("4.3", p, s, "2 files found.", "", "X", "2");

    this.t("4.4", p, s, "No files found.", "", "X", "0");
    this.t("4.5", p, s, "One file found.", "", "X", "1");
    this.t("4.6", p, s, "2 files found.", "", "X", "2");

    p.setLocale("ru");
    s = "{X_PLURAL:Найден {X:|один|{X}} файл|Найдены {X} файла|{X:Не найдено|Найдено {X}} файлов}.";

    this.t("5.1", p, s, "Не найдено файлов.", "", "X", "0");
    this.t("5.2", p, s, "Найден один файл.", "", "X", "1");
    this.t("5.3", p, s, "Найдены 2 файла.", "", "X", "2");
    this.t("5.4", p, s, "Найдено 5 файлов.", "", "X", "5");

    p.setLocale("en");
    s = "{FOO}";

    this.t("6.1", p, s, "1", "", "FOO", "1");
    this.t("6.2", p, s, "5.5", "", "FOO", "5.5");
    this.t("6.3", p, s, "bar", "", "FOO", "bar");
  }
}

class rendering {
  public static void main(String[] args) throws PlurrLocaleNotFoundException {
    RenderingTest test = new RenderingTest();
    test.run();
  }
}

File: python/t/rendering.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import sys
sys.path.append("..")

from plurr import Plurr


def t(n, p, s, params, options, message, exception=None):
    result = None
    try:
        result = p.format(s, params, options)
        if exception is not None:
            print("n: fail: should produce exception [{0}]".format(exception))
            return
        print(n + ": "  + ('pass' if result == message else "fail: [{0}] vs [{1}]".format(result, message)))
    except Exception as e:
        s = '{0}: {1}'.format(e.__class__.__name__, str(e))
        print(n + ": "  + ('pass' if s == exception else "fail: [{0}] vs [{1}]".format(s, exception)))


p = Plurr()

t('e.1', p, '', None, None, '', "TypeError: 'params' is not a dict")
t('e.2', p, 'err {', {}, None, '', "SyntaxError: Unmatched { found")
t('e.3', p, 'err }', {}, None, '', "SyntaxError: Unmatched } found")
t('e.4', p, '{foo}', {}, None, '', "LookupError: 'foo' not defined")
t('e.5', p, '{N_PLURAL}', {'N': 'NaN'}, None, '', "ValueError: Value of 'N' is not a zero or positive integer number")
t('e.6', p, '{N_PLURAL}', {'N': 1.5}, None, '', "ValueError: Value of 'N' is not a zero or positive integer number")

s = 'Do you want to delete {N_PLURAL:this {N} file|these {N} files} permanently?'

t('1.1', p, s, {'N': 1}, None, 'Do you want to delete this 1 file permanently?')
t('1.2', p, s, {'N': 5}, None, 'Do you want to delete these 5 files permanently?')

s = 'Do you want to drink {CHOICE:coffee|tea|juice}?'

t('2.1', p, s, {'CHOICE': 0}, None, 'Do you want to drink coffee?')
t('2.2', p, s, {'CHOICE': 1}, None, 'Do you want to drink tea?')
t('2.3', p, s, {'CHOICE': 2}, None, 'Do you want to drink juice?')

s = u'Удалить {N_PLURAL:этот {N} файл|эти {N} файла|эти {N} файлов} навсегда?'

t('3.1', p, s, {'N': 1}, {'locale': 'ru'}, u'Удалить этот 1 файл навсегда?')
t('3.2', p, s, {'N': 2}, {'locale': 'ru'}, u'Удалить эти 2 файла навсегда?')
t('3.3', p, s, {'N': 5}, {'locale': 'ru'}, u'Удалить эти 5 файлов навсегда?')
t('3.4', p, s, {'N': 5}, None, u'Удалить эти 5 файла навсегда?')

s = '{X_PLURAL:{X:|One|{X}} file|{X:No|{X}} files} found.'

t('4.1', p, s, {'X': 0}, None, 'No files found.')
t('4.2', p, s, {'X': 1}, None, 'One file found.')
t('4.3', p, s, {'X': 2}, None, '2 files found.')

t('4.4', p, s, {'X': '0'}, None, 'No files found.')
t('4.5', p, s, {'X': '1'}, None, 'One file found.')
t('4.6', p, s, {'X': '2'}, None, '2 files found.')

s = u'{X_PLURAL:Найден {X:|один|{X}} файл|Найдены {X} файла|{X:Не найдено|Найдено {X}} файлов}.'

t('5.1', p, s, {'X': 0}, {'locale': 'ru'}, u'Не найдено файлов.')
t('5.2', p, s, {'X': 1}, {'locale': 'ru'}, u'Найден один файл.')
t('5.3', p, s, {'X': 2}, {'locale': 'ru'}, u'Найдены 2 файла.')
t('5.4', p, s, {'X': 5}, {'locale': 'ru'}, u'Найдено 5 файлов.')

s = '{FOO}'

t('6.1', p, s, {'FOO': 1}, None, '1')
t('6.2', p, s, {'FOO': 5.5}, None, '5.5')
t('6.3', p, s, {'FOO': 'bar'}, None, 'bar')

s = u'Удалить {N_PLURAL:этот {N} файл|эти {N} файла|эти {N} файлов} навсегда?'
p.set_locale('ru')
t('7.1', p, s, {'N': 5}, None, u'Удалить эти 5 файлов навсегда?')
t('7.2', p, s, {'N': 5}, {'locale': 'en'}, u'Удалить эти 5 файла навсегда?')
t('7.3', p, s, {'N': 5}, None, u'Удалить эти 5 файлов навсегда?')
p.set_locale('en');
t('7.4', p, s, {'N': 5}, None, u'Удалить эти 5 файла навсегда?')

# Test for locale overrides and fallback
s = 'Test {N_PLURAL:Foo|Bar {N}}'
p = Plurr({'locale': 'ru'})
p._plural_equations['foo-bar'] = lambda n: 1 if (n != 2) else 0
t('8.1', p, s, {'N': 2}, {'locale': 'foo-bar'}, 'Test Foo')
t('8.2', p, s, {'N': 2}, {'locale': 'foo_BAR'}, 'Test Bar 2')
